### DISCUSSION TITLE
garbled chars with ctrl after exiting ssh session

### DISCUSSION THREAD

**Discussion by pdeva** [ghostty-org/ghostty#4015] [Q&A] [answered]
on mac os.
i ssh into a linux box and run zellij.
leave computer, return after few hours and wake it up from sleep. ghostty terminal shows ssh connection has been terminated.

now in that session, which is now on my local computer, whenever i use a key combination with `ctrl`, ghostty shows weird characters.
for example, this is me pressing `ctrl+c` twice:
<img width="1028" alt="Screenshot 2024-12-29 at 10 42 38 PM" src="https://github.com/user-attachments/assets/a38e8a37-7b2a-40e8-8c50-880d2f491a49" />

**Comment by dgl** [comment]
What software are you running on the remote system? (any software open or just a shell prompt? which shell? and same shell on local and remote?)

The likelihood is something has left a key reporting mode on. NeoVim for example can do this if it exits abnormally (you can reproduce this by a `kill -9` on nvim, but be careful!), potentially some shells and extensions(?) may turn this on too.

For zsh it is possible to workaround this by a snippet something like this in .zshrc which is a bit of hack, but does solve the problem. (Source of that is me via my openssh-unix-dev [list post](https://marc.info/?l=openssh-unix-dev&m=172040617312741&w=2), discussing how ssh itself could maybe handle this better.)

```zsh
function skip-csi-sequence() {
  local key
  while read -sk key && (( $((#key)) < 0x40 || $((#key)) > 0x7E )); do
    # empty body
  done
  if [[ $key = u ]] || [[ $key = "~" ]]; then
    # Ensure keys are sent normally, zle doesn't expect modifyOtherKeys / CSI u
    # (This means this key press is discarded, but the next one should be handled.)
    printf '\e[>4m\e[<u'
  fi
}

zle -N skip-csi-sequence
bindkey '\e[' skip-csi-sequence
```

You can also in many cases with problems like this type `reset` and enter (sometimes you need to do this without being able to see what you're typing).

**Reply by pdeva** [reply]
i run bash with zellij on the remote linux box. this issue doesn't happen with any other terminal like alacrity, warp etc. only with ghostty.

**Reply by pdeva** [reply]
running `reset` does indeed seem to solve it for the session tho.

**Comment by pdeva** [comment, marked as answer]
seems this was an issue with `atuin` and how it deals with kitty protocol, not with ghostty.
